test/Microsoft.AspNetCore.Mvc.Razor.Test/RazorPageTest.cs

```diff
@@ -1406,25 +1406,94 @@ public static TheoryData WriteAttributeData
             var suffix = string.Empty;
 
             // Act
-            page.BeginWriteAttributeTo(writer, "someattr", prefix, 0, suffix, 0, attributeValues.Length);
+            page.PushWriter(writer);
+            page.BeginWriteAttribute("someattr", prefix, 0, suffix, 0, attributeValues.Length);
             foreach (var value in attributeValues)
             {
-                page.WriteAttributeValueTo(
-                    writer,
+                page.WriteAttributeValue(
                     value.Item1,
                     value.Item2,
                     value.Item3,
                     value.Item4,
                     value.Item3?.ToString().Length ?? 0,
                     value.Item5);
             }
-            page.EndWriteAttributeTo(writer);
+            page.EndWriteAttribute();
+            page.PopWriter();
 
             // Assert
             Assert.Equal(expectedOutput, writer.ToString());
         }
 
         [Fact]
+        public void PushWriter_SetsUnderlyingWriter()
+        {
+            // Arrange
+            var page = CreatePage(p => { });
+            var writer = new StringWriter();
+
+            // Act
+            page.PushWriter(writer);
+
+            // Assert
+            Assert.Same(writer, page.ViewContext.Writer);
+        }
+
+        [Fact]
+        public void PopWriter_ResetsUnderlyingWriter()
+        {
+            // Arrange
+            var page = CreatePage(p => { });
+            var writer = new StringWriter();
+            page.PushWriter(writer);
+
```

Comment:
Grab the current writer (even if its null) here. `var expectedWriter = pge.ViewContext.Writer`. That way you can simply do an equals below.